### PR TITLE
Added support for long texture names

### PR DESCRIPTION
**Pull request by bmsq**
Fixes long texture name support in Doom Builder X.

- [x] Lump.MakeLongName()
- [x] Lump.MakeNames() 
- [x] Texture Loading
- [x] Search and Replace
- [x] Thing Sprites

### CODE REVIEW

File: Source/Core/IO/Lump.cs [hunk 1]
```diff
@@ -54,6 +54,9 @@ public class Lump
 		// Disposing
 		private bool isdisposed = false;
 
+        // bmsq - string to longname index, used when name.length > 8
+        private static Dictionary<string, long> longnameindex = new Dictionary<string, long>();
```
Comment (bmsq) [inline Source/Core/IO/Lump.cs]: I'm not overly happy with this living in the Lump class.  Ideally it should be in the DataManager but initialization order is a problem.   It should also be cleared each time a new map is loaded.  

This works as a proof of concept for now.

**Comment by bmsq** [comment]
Code styles for if statements are a little inconsistent throughout the doom builder code base.  Should I be using "if(expression)" or "if (expression)"?

**Comment by anotak** [comment]
looks good so far!
`if()` vs `if ()` is up to you about whatever feels appropriate at the moment. i personally tend to prefer `if()` but it's not a hard rule, sometimes the formatting of surrounding code lends itself to the `if ()` being more pleasing. the previous original Doom Builder 2 code (and the occasional bit of ported GZDB code) tends to be inconsistent about formatting, so there is no real system to any of it.

**Comment by bmsq** [comment]
I really wanted to add the LongNameIndex to DataManager as it seemed the most appropriate system to maintain this dictionary.  Unfortunately ResourceImages are created (which calls Lump.MakeLongName) in a number of awkward spots such as D3DDevice and Renderer2D, making this impossible without refactoring a lot of the map initialization code.

Instead, I've opted for the pragmatic approach of storing this information in the MapManager.  Its not ideal but its not technically wrong either.

Finally, I'm not convinced that Lump.MakeNames() needs updating.  It's a private function of Lump which is only ever called from the constructor.  Since its part of WAD loading, is the name restricted to 8 char based on the WAD format?

File: Source/Core/IO/Lump.cs [hunk 2]
```diff
@@ -81,11 +81,11 @@ internal Lump(Stream data, WAD owner, byte[] fixedname, int offset, int length)
 			this.offset = offset;
 			this.length = length;
 
-            // Make name
-            MakeNames(fixedname);
```
Comment (bmsq) [inline Source/Core/IO/Lump.cs:85]: Most changes in this file is just whitespace issues

**Comment by anotak** [comment]
good stuff.

the compromise with mapmanager is fine. the architecture of DB2 (that DBX has been based on) has often times forced me to make similar choices.

regarding MakeNames(): oh, yeah, you're right. i guess maybe a comment should be left by MakeNames() to make that more clear that it doesn't handle that case, just in case something changes in the future. but i guess i can merge this PR assuming you feel it is ready? i can leave the comment myself

**Comment by bmsq** [comment]
I haven't been able to test the sprite with long names yet.  

Does doom builder only read sprites from WAD files? I can't see how you configure it to load them from a directory.

**Comment by anotak** [comment]
i can't double check what you mean properly at the moment but if when you open a map, or set the map options in an already open map (default hotkey: F2), it gives an option to add a resource. there are 3 tabs in that window to let you select directories or pk3s. if that doesn't work, then perhaps it isn't implemented, i'll have to have a look later.

**Comment by bmsq** [comment]
I figured out what I was doing wrong. Sprite images need to be in a subdirectory called "sprites" in order for doombuilder to find them in a resource directory.

I'll test this today, hopefully it will be ready to merge by the weekend.

**Comment by bmsq** [comment]
All sprite name issues have been tested and resolved. 

I think this is ready to be merged.  Let me know if there is anything you're not happy with.

**Comment by bmsq** [comment]
Grrrr... I just discovered that sprite long names are created during configuration load.  This causes a null pointer when editing configurations before a map is loaded.  

What a pain, I'm going to have to think on this one.

**Comment by bmsq** [comment]
Sprite long name and initialization issues have been resolved.

I've updated Lump.MakeLongName() to guard against a scenario where General.Map has not been created.  This only happens when  user edits configurations before a map has been loaded/created.  In the event that this does occur, Lump.MakeLongName falls back to the original behavior. 

Having a truncated LongName isn't a problem until General.Map is created, at which point all LongNames are regenerated when resources are reloaded during map initialization.

Let me know you're not happy with this approach.  Otherwise this is ready to be merged.

**Comment by anotak** [comment]
looks good!
thanks!

**Pull request merged**